### PR TITLE
remove device shards

### PR DESCRIPTION
**Pull request by rongou**
Remove `DeviceShard` classes since we no longer support single process multi-GPU.

In `updater_gpu_hist.cu`, I only renamed the `DeviceShard` to `GPUHistMakerDevice` since it's fairly self contained. Let me know if you have a better name for it.

Hopefully the last cleanup PR for #4531.

@RAMitchell @trivialfis @sriramch

### CODE REVIEW

**Comment by trivialfis** [comment]
Sorry for the long wait.  Will give a proper review tomorrow.

File: src/common/hist_util.cu [hunk 1]
```diff
@@ -27,9 +27,11 @@ namespace common {
```
Comment (trivialfis) [inline src/common/hist_util.cu]: @RAMitchell is doing sketching algorithm anyway, these code is becoming legacy sooner or later.  It's not very motivating to change them ...

File: src/linear/updater_gpu_coordinate.cu [hunk 2]
```diff
@@ -235,29 +145,100 @@ class GPUCoordinateUpdater : public LinearUpdater {  // NOLINT
...
+        gpair_.data(),
+        host_gpair.data(),
+        gpair_.size() * sizeof(GradientPair), cudaMemcpyHostToDevice));
+  }
+
   // training parameter
   LinearTrainParam tparam_;
   CoordinateParam coord_param_;
-  int device_{};
+  int device_{-1};
```
Comment (trivialfis) [inline src/linear/updater_gpu_coordinate.cu]: We should use the `generic_parameter` in the future.

File: src/tree/updater_gpu_hist.cu [hunk 3]
```diff
@@ -1076,26 +1074,26 @@ class GPUHistMakerSpecialised {
...
       const DMatrix* data, HostDeviceVector<bst_float>* p_out_preds) {
-    if (shard_ == nullptr || p_last_fmat_ == nullptr || p_last_fmat_ != data) {
+    if (maker_ == nullptr || p_last_fmat_ == nullptr || p_last_fmat_ != data) {
       return false;
     }
     monitor_.StartCuda("UpdatePredictionCache");
     p_out_preds->SetDevice(device_);
-    dh::safe_cuda(cudaSetDevice(shard_->device_id));
-    shard_->UpdatePredictionCache(p_out_preds->DevicePointer());
+    dh::safe_cuda(cudaSetDevice(maker_->device_id));
```
Comment (trivialfis) [inline src/tree/updater_gpu_hist.cu]: Maybe in the future we can move this into `maker_`.

File: src/linear/updater_gpu_coordinate.cu (hunk 2 above)
Comment (rongou) [reply src/linear/updater_gpu_coordinate.cu]: Done.

File: src/tree/updater_gpu_hist.cu (hunk 3 above)
Comment (rongou) [reply src/tree/updater_gpu_hist.cu]: Done.

**Comment by trivialfis** [comment]
Merging as so we can have more time to test it.

**Pull request merged**